modules/node-shared/src/main/scala/io/constellationnetwork/node/shared/infrastructure/snapshot/CurrencySnapshotAcceptanceManager.scala

```diff
@@ -492,50 +492,50 @@ object CurrencySnapshotAcceptanceManager {
-            SortedMap.empty[Address, List[SpendAction]].pure[F]
+            empty
           } else {
-            val snapshotsMap = lastGlobalSnapshots.map { snapshots =>
-              snapshots.map(s => (s.ordinal, s)).toMap
-            }.getOrElse(Map.empty)
-
-            val start = lastSyncOrdinal.value.value
-            val end = lastGlobalSnapshot.ordinal.value.value
-            val snapshotOrdinals = (start until end)
-              .map(l => SnapshotOrdinal(NonNegLong.unsafeFrom(l)))
-              .toList
-
-            val limitOfOrdinalsToFetch = 20L
-            if (snapshotOrdinals.length > limitOfOrdinalsToFetch) {
+            val snapshotCache: Map[SnapshotOrdinal, Hashed[GlobalIncrementalSnapshot]] =
+              lastGlobalSnapshots.map(_.map(s => s.ordinal -> s).toMap).getOrElse(Map.empty)
+
+            val startOrdinal = lastSyncOrdinal.value.value
+            val endOrdinal = lastGlobalSnapshot.ordinal.value.value
+            val snapshotOrdinals: List[SnapshotOrdinal] =
+              (startOrdinal until endOrdinal).map(i => SnapshotOrdinal(NonNegLong.unsafeFrom(i))).toList
+
+            val maxAllowedGap = 20
+            if (snapshotOrdinals.size > maxAllowedGap) {
               Slf4jLogger
                 .getLogger[F]
                 .warn(
-                  s"Interval of ordinals of metagraph $currencyId between lastSyncGlobalSnapshot and lastGlobalView greater than $limitOfOrdinalsToFetch, skipping fetching interval"
-                )
-                .as(
-                  lastGlobalSnapshot.spendActions.getOrElse(SortedMap.empty[Address, List[SpendAction]])
+                  s"Interval of ordinals of metagraph $currencyId between lastSyncGlobalSnapshot and lastGlobalView is greater than $maxAllowedGap; skipping fetching interval"
                 )
+                .as(lastGlobalSnapshot.spendActions.getOrElse(SortedMap.empty))
             } else {
-              snapshotOrdinals.foldMapM { ordinal =>
-                snapshotsMap.get(ordinal) match {
-                  case Some(snapshot) =>
-                    snapshot.spendActions.getOrElse(SortedMap.empty[Address, List[SpendAction]]).pure[F]
-                  case None =>
-                    getGlobalSnapshotByOrdinal(ordinal).map(
-                      _.flatMap(_.spendActions)
-                        .getOrElse(SortedMap.empty[Address, List[SpendAction]])
-                    )
+              val (cached, missing) = snapshotOrdinals.partition(snapshotCache.contains)
+
+              val fromCache: List[SortedMap[Address, List[SpendAction]]] =
+                cached.flatMap(ordinal => snapshotCache.get(ordinal).flatMap(_.spendActions).toList)
+
+              val fetchMissing: F[List[SortedMap[Address, List[SpendAction]]]] =
+                missing.parTraverse { ordinal =>
+                  getGlobalSnapshotByOrdinal(ordinal)
+                    .map(_.flatMap(_.spendActions).getOrElse(SortedMap.empty))
```

Comment:
If one of the missing values errors out while trying to retrieve it, then should this whole operation be skipped or is should just that element be skipped? I'm just thinking that this returns an `F[_]` wrapped value which might error so can we handle the case of missing one or do we need the guarantee that all elements have been fetched?